rust/kernel/thread.rs

```diff
@@ -0,0 +1,198 @@
+    /// This function might sleep due to the memory allocation and waiting for completion in
+    /// `kthread_create_on_node`. Therefore cannot call this in a atomic context.
+    ///
+    /// # Examples:
+    /// use kernel::thread::Thread;
+    /// use alloc::boxed::Box;
+    ///
+    /// let mut a = 1;
+    ///
+    /// let t = Thread::new(
+    ///   move || {
+    ///     let b = Box::try_new(42)?;
+    ///
+    ///     for _ in 0..10 {
+    ///         a = a + 1;
+    ///         println!("Hello Rust Thread {}", a + b.as_ref());
+    ///     }
+    ///     Ok(())
+    ///   },
+    ///   cstr!("rust-thread")
+    /// )?;
+    ///
+    /// t.wake_up();
+    pub fn new<F>(f: F, name: CStr) -> KernelResult<Self>
+    where
+        F: FnOnce() -> KernelResult<()>,
+        F: Send + 'static,
+    {
+        // Allocate closure here, because this function maybe returns before `rust_thread_func`
+        // (the function that use the closure) get executed.
+        let bf: Box<dyn FnOnce() -> KernelResult<()> + 'static> = Box::try_new(f)?;
+
+        // Double boxing here because `dyn FnOnce` is a fat pointer, and we can only pass a usize
+        // as the `data` for `kthread_create_on_node`.
+        //
+        // We `into_raw` from this side, and will `from_raw` at the other side to transfer the
+        // ownership of the boxed data.
+        let db = Box::into_raw(Box::try_new(bf)?) as *mut _;
+
+        let task;
+
+        // SAFETY:
+        //
+        // * `db` is a proper pointer (generated by `Box::into_raw()`), and if succeed, the new
+        // thread will get the ownership.
+        //
+        // * `kthread_create_on_node` will copy the content of name, so we don't need to make the
+        // name live longer.
+        unsafe {
+            task = bindings::kthread_create_on_node(
+                Some(rust_thread_func),
+                db,
+                bindings::NUMA_NO_NODE,
+                "%s".as_ptr() as _,
+                name.as_ptr(),
+            );
+        }
+
+        let result = ptr_to_result(task);
+
+        if result.is_err() {
+            // Creation fails, we need to get back the double boxed closure.
+            //
+            // SAFETY:
+            //
+            // `db` is a proper pointer generated by a `Box::into_raw()` from a box created by us,
+            // if the thread creation fails, no one will consume that pointer.
+            unsafe {
+                Box::from_raw(db);
+            }
+        } else {
+            // Increase the refcount of the task, so that it won't go away if it `do_exit`.
+            //
+            // SAFETY:
+            //
+            // `task` is a valid pointer to newly created thread
+            unsafe {
+                rust_helper_get_task_struct(task);
+            }
+        }
+
+        Ok(Thread { task: result? })
+    }
+
+    /// Wake up the kernel thread, note that a newly created kthread will not run until `wake_up`
+    /// is called.
```

Comment:
The first paragraph should be a single sentence that summarizes the function. The rest of details should go in different paragraphs.

i.e. similar to Git commits and their first line.